### PR TITLE
[IOPAY-29] Polling refactoring

### PR DESCRIPTION
**Pull request by infantesimone**
#### Description

This PR introduces a polling refactoring in order to take into account the transaction final status.

#### List of Changes

1. Polling refactoring, in particular the transient error condition to consider transaction final status.
2. e2e test  with `Puppeteer` completing a 3ds payment;
 
#### Motivation and Context

It is important  to take into account the final state in the polling conditions as during the _Method_ or _Challenge_ 3ds steps they can verify errors and therefore lead to a final state of error.

#### How Has This Been Tested?

- `yarn test:docker-int pay3ds2Workflow.browser.integration.docker.test` (with local [Payment Manager](https://github.com/pagopa/pm-mock))

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/__integrations__/docker/pay3ds2Workflow.browser.integration.docker.test.ts [hunk 1]
```diff
@@ -114,4 +120,124 @@ describe('IOPAY App', () => {
...
+      ),
+      page.click(payButtonS),
+      page.waitForNavigation(),
+    ]);
+
+    expect(pay3ds2Response.status()).toEqual(200);
+
+    // polling method Step and Acs redirect
+    // eslint-disable-next-line functional/no-let
+    let waitForMethodUrl = true;
```
Comment (pasqualespica) [inline src/__integrations__/docker/pay3ds2Workflow.browser.integration.docker.test.ts:195]: great : ) . I used this for mixpanel test events too

File: src/response.ts [hunk 2]
```diff
@@ -127,25 +141,32 @@ document.addEventListener('DOMContentLoaded', async () => {
...
-              start3DS2AcsChallengeStep(transactionStatus.data.acsUrl, transactionStatus.data.params, document.body),
-          )
-          .run();
-      }
+      await fromPredicate<Error, MessageEvent<any>>(
+        e1 => e1.origin === 'http://localhost:7071' && e1.data === '3DS.Notification.Received',
+        toError,
+      )(e)
+        .fold(
+          _ => null, // TODO error handle
```
Comment (pasqualespica) [inline src/response.ts:149]: When we'll merge this [PR](https://github.com/pagopa/io-pay/pull/49) remember to put here `THREEDSMETHODURL_STEP1_RESP_ERR`

File: src/utils/testUtils.ts [hunk 3]
```diff
@@ -18,3 +18,23 @@ export async function getIdPayment(pmControlHost: string, pmControlPort: string)
   await myBrowser.close();
   return idPayment;
 }
+
+export async function setMethodChallengeSteps(pmControlHost: string, pmControlPort: string) {
```
Comment (pasqualespica) [inline src/utils/testUtils.ts:22]: great very useful

**Review by pasqualespica — APPROVED** [review]
Great 👍 Now with this PR we have all 3ds2 payment flow tested : )

**Review by GiovanniMancini — APPROVED** [review]

**Pull request merged**